### DISCUSSION TITLE
Yet another serialization format?

### DISCUSSION THREAD

**Discussion by casey** [rust-bitcoin/rust-bitcoin#1087] [Ideas]
I've worked on a few Bitcoin-adjacent projects where I've needed a general purpose serialization format, but haven't been able to use an existing one because I haven't been able to find one that met the requirements of the project.

I feel like those requirements are actually pretty common for Bitcoin-adjacent projects, so I wanted to float the idea of writing a new one, and see if there was any interest.

The requirements that I have, and that I suspect other Bitcoin-ecosystem projects have, which seem to be mostly unmet by existing formats are:

- Very simple, to minimize the chance of bugs or security vulnerabilities
- Reasonably efficient
- Efficient support of binary data, i.e., byte arrays
- Canonical: Every value has exactly one serialized form that serializers can produce and deserializers can consume
- Easy to extract sub-messages as byte sequences so that they can be hashed, so that they can be signed and so signatures over them can be verified, and so that they can be referred to by hash, e.g., in consensus critical contexts.
- General purpose, i.e., doesn't force me to define custom encodings for my types
- Schema evolution, i.e., new fields can be added to messages and new variants can be added to enums in a backwards-compatible way

And one thing I'm undecided about:
- Self-describing, e.g. like JSON and not like protobufs, so that a parser can decode a message and render it in a text-based human readable format, to make debugging easy. Increases complexity a little, but debugging non-self-describing binary formats can be extremely painful, so may be worth it.

I've looked at a bunch of different formats, but none of them fit the bill. Some formats that I've looked at:

- CBOR:
  - Complex enough that different implementations likely support different subsets of the full specification
  - Values can be encoded in many different ways, enforcing canonicity is challenging

- bencode, the serialization format used in `.torrent` files:
  - Actually pretty good
  - Has well defined canonicity rules
  - Some details are weird, like encoding numbers as decimal ASCII
  - Only binary strings are allowed as dictionary keys, which increases their size
  - Obscure

- JSON:
  - Inefficient for binary data
  - A few details of the format, like the fact that it's text based and can use escape sequences in strings, makes canonicity challenging 

I have some ideas for a simple serialization format that would meet all the above requirements, but before I open the doors to that discussion and all the bike-shedding that entails, I wanted to see what people thought:

- Have you wanted a similar serialization format?
- Do you think some of the requirements are unnecessary?
- Do you have additional requirements?
- Are there good options that should be considered?
- Is an entirely new serialization format warranted, or should we instead create a well-defined, simple subset of CBOR with canonicity rules?

**Comment by DanGould** [comment]
I think [Cap'n Proto ](https://capnproto.org/) satisfies all of your requirements

**Reply by casey** [reply]
My main hesitation is that Cap'n Proto seems pretty complicated. I do notice that it was somewhat recently pulled into Bitcoin Core. Does anyone have a link to the thread/issue where that was discussed? I'm really curious about the rationale, and wonder if some of the issues above were discussed anywhere.

**Comment by Kixunil** [comment]
Was thinking about it but Cap'n Proto looks good enough. One thing it (or the Rust impl) specifically does *not* support is newtypes. Also if I remember correctly it has quite high MSRV.

**Reply by casey** [reply]
Cap'n proto is good, but I think it fails the simplicity test. It requires an external schema, a parser to parse that schema, is very complicated, doesn't canonicalize by default, and the rust implementation uses a bunch of `unsafe`. Also, it's unclear to me how to extract sub-messages for hasing.

**Comment by dpc** [comment]
Whenever I evaluate serialization, I usually end up with `msgpack`. BTW. Are you aware of:

https://github.com/djkoloski/rust_serialization_benchmark

?

**Reply by casey** [reply]
I'll have to check out the spec, I know of msgpack, but I haven't gotten into the details.

That benchmarking page is awesome, but speed is a very peripheral consideration.

**Reply by dpc** [reply]
> That benchmarking page is awesome, but speed is a very peripheral consideration.

I think the list of all formats is more valuable than the metrics itself. :D

**Comment by Kixunil** [comment]
Remembered @dr-orlovsky designed some serialization for RGB using only a bunch of traits and helpers. You might want to check it out since it's cryptography application. Not sure about the exact details related to your requirements.

**Reply by casey** [reply]
Good lead! I checked it out and it seems promising, but there are no specs, so I couldn't figure out things like how structs/enums are encoded, without reading a bunch of quite complex derive macro code.

**Reply by dr-orlovsky** [reply]
I assume you are talking about https://crates.io/crates/strict_encoding

Well, it is designed specifically for client-side-validation needs; i.e. it has some embedded restrictions and deviates from bitcoin consensus encoding things like `VarInt` types. By default, it restricts the size of all arrays, maps, sets to 64k element (16-bit length). You can do larger structures by using special types `MediumVec`, `LargeVec` (24 bits and 32 bits). With these I use the encoding not only in RGB, but also for networking communications. It is not that efficient though as Cap'n'Proto in terms of performance - but simplifies many other things.

The docs are still pending; right now we have only https://docs.rs/strict_encoding/0.8.1/strict_encoding/index.html and a dedicated doc website is WIP at https://strict-encoding.lnp-bp.org/. However, the overall use is quite simple: just add `#[derive(StrictEncode, StrictDecode)]` to any structure - and that's all; it will be encodable in the same way as with bitcoin consensus encoding (just use `T::strict_encode/decode` or `T::strict_serialize/deserialize` methods). There are some specifics in working with enums, and they are described with samples in https://docs.rs/strict_encoding_derive/0.8.0/strict_encoding_derive/index.html.

**Reply by casey** [reply]
Thanks for the info! How does strict encoding handle message and enum extensibility? I.e. can you add new messages and fields to enums without breaking old clients?

**Reply by dr-orlovsky** [reply]
You can, in two ways:

1. With structs, you use TLV extensions which work exactly like TLVs in LN bolts. Strict encoding + TLVs is called "network encoding" and derived with `#[derive(NetworkEncode, NetworkDecode)]` and used like `#[network_encoding(tlv = ...]`. There are examples in the https://docs.rs/strict_encoding_derive/0.8.0/strict_encoding_derive/index.html.

2. With enums, you can only extend enums defining list of API messages. This is a different topic from strict encoding and is a part of special `internet2` crate which allows creation LN-style P2P and RPC APIs. You can check how it works here: https://github.com/Internet2-WG/rust-internet2/blob/master/derive/examples/peer_api.rs and https://github.com/Internet2-WG/rust-internet2/blob/master/derive/tests/api_strict.rs. You can read more on the related projects in https://github.com/Internet2-WG

Please note that this works like in LN with even/odd rule and unknown TLV fields.
